external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,454 @@
+
+                    kafkaSpoutStreams.emit(collector, messageId);           // emits one tuple per record
+                    LOG.debug("Emitted tuple [{}] for record [{}]", tuple, record);
+                }
+            }
+        }
+    }
+
+    private void commitOffsetsForAckedTuples() {
+        // Find offsets that are ready to be committed for every topic partition
+        final Map<TopicPartition, OffsetAndMetadata> nextCommitOffsets = new HashMap<>();
+        for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+            final OffsetAndMetadata nextCommitOffset = tpOffset.getValue().findNextCommitOffset();
+            if (nextCommitOffset != null) {
+                nextCommitOffsets.put(tpOffset.getKey(), nextCommitOffset);
+            }
+        }
+
+        // Commit offsets that are ready to be committed for every topic partition
+        if (!nextCommitOffsets.isEmpty()) {
+            kafkaConsumer.commitSync(nextCommitOffsets);
+            LOG.debug("Offsets successfully committed to Kafka [{}]", nextCommitOffsets);
+            // Instead of iterating again, it would be possible to commit and update the state for each TopicPartition
+            // in the prior loop, but the multiple network calls should be more expensive than iterating twice over a small loop
+            for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+                final OffsetEntry offsetEntry = tpOffset.getValue();
+                offsetEntry.commit(nextCommitOffsets.get(tpOffset.getKey()));
+            }
+        } else {
+            LOG.trace("No offsets to commit. {}", toString());
+        }
+    }
+
+    // ======== Ack =======
+    @Override
+    public void ack(Object messageId) {
+        if (!consumerAutoCommitMode) {  // Only need to keep track of acked tuples if commits are not done automatically
+            final MessageId msgId = (MessageId) messageId;
+            acked.get(msgId.getTopicPartition()).add(msgId);
+            LOG.debug("Added acked message [{}] to list of messages to be committed to Kafka", msgId);
+        }
+    }
+
+    // ======== Fail =======
+
+    @Override
+    public void fail(Object messageId) {
+        final MessageId msgId = (MessageId) messageId;
+        if (msgId.numFails() < maxRetries) {
+            msgId.incrementNumFails();
+            kafkaSpoutStreams.emit(collector, msgId);
+            LOG.debug("Retried tuple with message id [{}]", msgId);
+        } else { // limit to max number of retries
+            LOG.debug("Reached maximum number of retries. Message [{}] being marked as acked.", msgId);
+            ack(msgId);
+        }
+    }
+
+    // ======== Activate / Deactivate / Close / Declare Outputs =======
+
+    @Override
+    public void activate() {
+        // Shouldn't have to do anything for now. If specific cases need to be handled logic will go here
```

Comment:
activate gets called after deactivate is called.  It can potentially go back and forth between the two states multiple times.  If we shutdown in deactivate, which I think is fine, then we need to reinitialize in activate.

I assume that once the kafkaConsumer is closed poll will not work on it any more.